File: source/framework/model/test/simulate_test.cpp
#define BOOST_TEST_MODULE lue framework model simulate
#include "lue/framework/model/model.hpp"
#include "lue/framework/model/progressor.hpp"
#include "lue/framework/model/simulate.hpp"
#include "lue/framework/test/hpx_unit_test.hpp"
// #include <hpx/config.hpp>
// #include <boost/test/unit_test.hpp>


namespace {

    class MyModel: public lue::Model
    {

        public:

            void preprocess([[maybe_unused]] lue::Count const sample_nr) final
            {
                ++preprocess_called;
            }


            void initialize() final
            {
                ++initialize_called;
            }


            hpx::shared_future<void> simulate([[maybe_unused]] lue::Count const time_step) final
            {
                ++simulate_called;

                return hpx::make_ready_future<void>();
            }


            void finalize() final
            {
                ++finalize_called;
            }


            void postprocess() final
            {
                ++postprocess_called;
            }


            lue::Count preprocess_called{0};
            lue::Count initialize_called{0};
            lue::Count simulate_called{0};
            lue::Count finalize_called{0};
            lue::Count postprocess_called{0};
    };


    class MyProgressor: public lue::Progressor
    {

        public:

            void preprocess([[maybe_unused]] lue::Count const sample_nr) final
            {
                ++preprocess_called;
            }


            void initialize() final
            {
                ++initialize_called;
            }


            void simulate([[maybe_unused]] lue::Count const time_step) final
            {
                ++simulate_called;
            }


            void finalize() final
            {
                ++finalize_called;
            }


            void postprocess() final
            {
                ++postprocess_called;
            }


            lue::Count preprocess_called{0};
            lue::Count initialize_called{0};
            lue::Count simulate_called{0};
            lue::Count finalize_called{0};
            lue::Count postprocess_called{0};
    };

}  // Anonymous namespace


BOOST_AUTO_TEST_CASE(run_deterministic)
{
    // Static model
    {
        MyModel model{};
        MyProgressor progressor{};
        lue::Count const nr_time_steps = 0;

        lue::run_deterministic(model, progressor, nr_time_steps);

        BOOST_CHECK_EQUAL(model.preprocess_called, 0);
        BOOST_CHECK_EQUAL(model.initialize_called, 1);
        BOOST_CHECK_EQUAL(model.simulate_called, nr_time_steps);
        BOOST_CHECK_EQUAL(model.finalize_called, 1);
        BOOST_CHECK_EQUAL(model.postprocess_called, 0);

        BOOST_CHECK_EQUAL(progressor.preprocess_called, 0);
        BOOST_CHECK_EQUAL(progressor.initialize_called, 1);
        BOOST_CHECK_EQUAL(progressor.simulate_called, nr_time_steps);
        BOOST_CHECK_EQUAL(progressor.finalize_called, 1);
        BOOST_CHECK_EQUAL(progressor.postprocess_called, 0);
    }

    // Dynamic model
    {
        MyModel model{};
        MyProgressor progressor{};
        lue::Count const nr_time_steps = 5;

        lue::run_deterministic(model, progressor, nr_time_steps);

        BOOST_CHECK_EQUAL(model.preprocess_called, 0);
        BOOST_CHECK_EQUAL(model.initialize_called, 1);
        BOOST_CHECK_EQUAL(model.simulate_called, nr_time_steps);
        BOOST_CHECK_EQUAL(model.finalize_called, 1);
        BOOST_CHECK_EQUAL(model.postprocess_called, 0);

        BOOST_CHECK_EQUAL(progressor.preprocess_called, 0);
        BOOST_CHECK_EQUAL(progressor.initialize_called, 1);
        BOOST_CHECK_EQUAL(progressor.simulate_called, nr_time_steps);
        BOOST_CHECK_EQUAL(progressor.finalize_called, 1);
        BOOST_CHECK_EQUAL(progressor.postprocess_called, 0);
    }
}


BOOST_AUTO_TEST_CASE(run_stochastic)
{
    // Static model
    {
        MyModel model{};
        MyProgressor progressor{};
        lue::Count const nr_samples = 5;
        lue::Count const nr_time_steps = 0;

        lue::run_stochastic(model, progressor, nr_samples, nr_time_steps);

        BOOST_CHECK_EQUAL(model.preprocess_called, nr_samples);
        BOOST_CHECK_EQUAL(model.initialize_called, nr_samples);
        BOOST_CHECK_EQUAL(model.simulate_called, nr_samples * nr_time_steps);
        BOOST_CHECK_EQUAL(model.finalize_called, nr_samples);
        BOOST_CHECK_EQUAL(model.postprocess_called, nr_samples);

        BOOST_CHECK_EQUAL(progressor.preprocess_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.initialize_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.simulate_called, nr_samples * nr_time_steps);
        BOOST_CHECK_EQUAL(progressor.finalize_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.postprocess_called, nr_samples);
    }

    // Dynamic model
    {
        MyModel model{};
        MyProgressor progressor{};
        lue::Count const nr_samples = 5;
        lue::Count const nr_time_steps = 5;

        lue::run_stochastic(model, progressor, nr_samples, nr_time_steps);

        BOOST_CHECK_EQUAL(model.preprocess_called, nr_samples);
        BOOST_CHECK_EQUAL(model.initialize_called, nr_samples);
        BOOST_CHECK_EQUAL(model.simulate_called, nr_samples * nr_time_steps);
        BOOST_CHECK_EQUAL(model.finalize_called, nr_samples);
        BOOST_CHECK_EQUAL(model.postprocess_called, nr_samples);

        BOOST_CHECK_EQUAL(progressor.preprocess_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.initialize_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.simulate_called, nr_samples * nr_time_steps);
        BOOST_CHECK_EQUAL(progressor.finalize_called, nr_samples);
        BOOST_CHECK_EQUAL(progressor.postprocess_called, nr_samples);
    }
}

File: source/framework/python/test/simulate_test.py
import time

import numpy as np

import lue.framework as lfr
import lue_test


def setUpModule():
    lue_test.start_hpx_runtime()


def tearDownModule():
    lue_test.stop_hpx_runtime()


class MyModel(lfr.Model):
    def __init__(self):
        lfr.Model.__init__(self)
        self.preprocess_called = 0
        self.initialize_called = 0
        self.simulate_called = 0
        self.finalize_called = 0
        self.postprocess_called = 0

    def preprocess(self, sample_nr):
        self.preprocess_called += 1

    def initialize(self):
        self.initialize_called += 1

    def simulate(self, time_step):
        # This sleep is relevant. Otherwise things go bad in the simulate function.
        # Possibly tasks get shuffled(?).
        time.sleep(0.1)
        self.simulate_called += 1

    def finalize(self):
        self.finalize_called += 1

    def postprocess(self):
        self.postprocess_called += 1


class MyProgressor(lfr.Progressor):
    def __init__(self):
        lfr.Progressor.__init__(self)
        self.preprocess_called = 0
        self.initialize_called = 0
        self.simulate_called = 0
        self.finalize_called = 0
        self.postprocess_called = 0

    def preprocess(self, sample_nr):
        self.preprocess_called += 1

    def initialize(self):
        self.initialize_called += 1

    def simulate(self, time_step):
        self.simulate_called += 1

    def finalize(self):
        self.finalize_called += 1

    def postprocess(self):
        self.postprocess_called += 1


class SimulateTest(lue_test.TestCase):
    @lue_test.framework_test_case
    def test_run_deterministic(self):
        model = MyModel()
        progressor = MyProgressor()
        nr_time_steps = 5

        lfr.run_deterministic(model, progressor, nr_time_steps, rate_limit=2)

        self.assertEqual(model.preprocess_called, 0)
        self.assertEqual(model.initialize_called, 1)
        self.assertEqual(model.simulate_called, nr_time_steps)
        self.assertEqual(model.finalize_called, 1)
        self.assertEqual(model.postprocess_called, 0)

        self.assertEqual(progressor.preprocess_called, 0)
        self.assertEqual(progressor.initialize_called, 1)
        self.assertEqual(progressor.simulate_called, nr_time_steps)
        self.assertEqual(progressor.finalize_called, 1)
        self.assertEqual(progressor.postprocess_called, 0)

    @lue_test.framework_test_case
    def test_run_stochastic(self):
        model = MyModel()
        progressor = MyProgressor()
        nr_samples = 3
        nr_time_steps = 5

        lfr.run_stochastic(model, progressor, nr_samples, nr_time_steps)

        self.assertEqual(model.preprocess_called, nr_samples)
        self.assertEqual(model.initialize_called, nr_samples)
        self.assertEqual(model.simulate_called, nr_samples * nr_time_steps)
        self.assertEqual(model.finalize_called, nr_samples)
        self.assertEqual(model.postprocess_called, nr_samples)

        self.assertEqual(progressor.preprocess_called, nr_samples)
        self.assertEqual(progressor.initialize_called, nr_samples)
        self.assertEqual(progressor.simulate_called, nr_samples * nr_time_steps)
        self.assertEqual(progressor.finalize_called, nr_samples)
        self.assertEqual(progressor.postprocess_called, nr_samples)
